### PR TITLE
backport: merge bitcoin#21404, #21415, #19314, #21882, #16939, #19084, #20602, #20253, #20480, #20736 (remove deprecated logic, use standard library alternatives)

### PR DESCRIPTION
**Pull request by kittywhiskers**

### CODE REVIEW

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Comment by PastaPastaPasta** [comment]
this conflicts with https://github.com/dashpay/dash/pull/5051, but trivially rebased on top of it. Going to review 5051 first

**Comment by PastaPastaPasta** [comment]
I've rebased this on top of develop, and resolved a trivial include conflict

**Review by UdjinM6 — APPROVED** [review]
utACK

File: src/evo/cbtx.cpp [hunk 1]
```diff
@@ -229,7 +229,7 @@ bool CalcCbTxMerkleRootQuorums(const CBlock& block, const CBlockIndex* pindexPre
     int64_t nTime1 = GetTimeMicros();
 
     auto retVal = CachedGetQcHashesQcIndexedHashes(pindexPrev, quorum_block_processor);
-    if (retVal == std::nullopt) {
+    if (!retVal) {
```
Comment (PastaPastaPasta) [inline src/evo/cbtx.cpp:232]: why?

File: src/llmq/utils.h [hunk 2]
```diff
@@ -11,9 +11,10 @@
 #include <random.h>
 #include <set>
 #include <sync.h>
+#include <versionbits.h>
```
Comment (PastaPastaPasta) [inline src/llmq/utils.h:14]: fine, but unrelated

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merge via merge commit

I have confirmed CI is happy

File: src/llmq/snapshot.cpp [hunk 3]
```diff
@@ -291,8 +291,8 @@ bool BuildQuorumRotationInfo(const CGetQuorumRotationInfo& request, CQuorumRotat
         response.mnListDiffAtHMinus4C = std::move(mn4c);
     } else {
         response.extraShare = false;
-        response.quorumSnapshotAtHMinus4C = std::nullopt;
-        response.mnListDiffAtHMinus4C = std::nullopt;
+        response.quorumSnapshotAtHMinus4C.reset();
+        response.mnListDiffAtHMinus4C.reset();
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/llmq/snapshot.cpp:295]: I personally think I disagree with both of these styling changes but it's fine

**Pull request merged**